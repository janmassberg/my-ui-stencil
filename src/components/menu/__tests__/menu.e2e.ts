import { newE2EPage } from "@stencil/core/testing";

const testMenuItemsFlat = [
  {
    name: "item-1",
    label: "Menu Item 1",
  },
  {
    name: "item-2",
    label: "Menu Item 2",
  },
  {
    name: "item-3",
    label: "Menu Item 3",
  },
];

const testMenuItemsNested = [
  {
    name: "item-1",
    label: "Menu Item 1",
    items: [
      {
        name: "item-1-1",
        label: "Menu Item 1-1",
        items: [
          {
            name: "item-1-1-1",
            label: "Menu Item 1-1-1",
          },
          {
            name: "item-1-1-2",
            label: "Menu Item 1-1-2",
          },
        ],
      },
    ],
  },
];

describe("ui-menu", () => {
  it("should handle mouse and keyboard events correctly", async () => {
    const page = await newE2EPage();
    await page.setContent(`
            <ui-menu name="test-menu"></ui-menu>`);
    const uiMenuChange = await page.spyOnEvent("uiMenuChange");

    const menu = await page.find("ui-menu.hydrated");
    menu.setProperty("items", testMenuItemsFlat);
    menu.setProperty("current", testMenuItemsFlat[0].name);
    await page.waitForChanges();

    const item2 = await page.find("ui-menu-item[name=item-2].hydrated");
    const item3 = await page.find("ui-menu-item[name=item-3].hydrated");

    // Initial current value
    expect(menu.getAttribute("current")).toEqual("item-1");

    // Mouse events
    await item2.click();
    await expect(uiMenuChange).toHaveReceivedEventDetail({
      name: "test-menu",
      current: "item-2",
    });
    await page.waitForChanges();
    // expect(menu.getAttribute('current')).toEqual('item-2');

    // Keyboard events
    await item2.press("ArrowRight");
    await item3.press("Enter");
    expect(uiMenuChange).toHaveReceivedEventDetail({
      name: "test-menu",
      current: "item-3",
    });
    await page.waitForChanges();
    // expect(menu.getAttribute('current')).toEqual('item-3');

    expect(menu).toMatchSnapshot();
  });

  it("should render nested menu items correctly", async () => {
    const page = await newE2EPage();
    await page.setContent(`
            <ui-menu name="test-menu"></ui-menu>`);
    const uiMenuChange = await page.spyOnEvent("uiMenuChange");
    const menu = await page.find("ui-menu.hydrated");

    menu.setProperty("items", testMenuItemsNested);
    await page.waitForChanges();

    const rootItem = await page.find("ui-menu-item[name=item-1].hydrated");
    let nestedItem;
    // expand level 1
    await rootItem.click();
    await page.waitForChanges();
    // expand level 2
    nestedItem = await page.find("ui-menu-item[name=item-1-1].hydrated");
    await nestedItem.click();
    await page.waitForChanges();

    // select nested item
    nestedItem = await page.find("ui-menu-item[name=item-1-1-2].hydrated");
    await nestedItem.click();
    await page.waitForChanges();

    await expect(uiMenuChange).toHaveReceivedEventDetail({
      name: "test-menu",
      current: "item-1-1-2",
    });

    await page.waitForChanges();
    expect(menu.getAttribute("current")).toEqual("item-1-1-2");
    expect(rootItem.getProperty("selected")).toBeTruthy();

    expect(menu).toMatchSnapshot();
  });
});
