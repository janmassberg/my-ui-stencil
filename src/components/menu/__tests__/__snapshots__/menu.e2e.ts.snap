// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui-menu should handle mouse and keyboard events correctly 1`] = `
<ui-menu class="hydrated ui-menu ui-menu--horizontal ui-menu--primary" current="item-3" name="test-menu" role="menubar">
  <!---->
  <div class="ui-menu__inner">
    <ui-menu-item aria-label="Menu Item 1" class="hydrated ui-menu-item" level="0" name="item-1" role="menuitem" tabindex="-1">
      <!---->
      <div class="ui-menu-item__item">
        <span class="ui-menu-item__label">
          <span class="ui-menu-item__label-slot"></span>
          <span class="ui-menu-item__label-text">
            Menu Item 1
          </span>
        </span>
      </div>
    </ui-menu-item>
    <ui-menu-item aria-label="Menu Item 2" class="hydrated ui-menu-item" level="0" name="item-2" role="menuitem" tabindex="-1">
      <!---->
      <div class="ui-menu-item__item">
        <span class="ui-menu-item__label">
          <span class="ui-menu-item__label-slot"></span>
          <span class="ui-menu-item__label-text">
            Menu Item 2
          </span>
        </span>
      </div>
    </ui-menu-item>
    <ui-menu-item aria-label="Menu Item 3" aria-selected="" class="hydrated ui-menu-item ui-menu-item--selected" level="0" name="item-3" role="menuitem" selected="" tabindex="-1">
      <!---->
      <div class="ui-menu-item__item">
        <span class="ui-menu-item__label">
          <span class="ui-menu-item__label-slot"></span>
          <span class="ui-menu-item__label-text">
            Menu Item 3
          </span>
        </span>
      </div>
    </ui-menu-item>
  </div>
  <ui-menu-indicator class="hydrated ui-menu-indicator ui-menu-indicator--horizontal" style="display: block; transform: translateX(246px); width: 122px;"></ui-menu-indicator>
</ui-menu>
`;

exports[`ui-menu should render nested menu items correctly 1`] = `
<ui-menu class="hydrated ui-menu ui-menu--horizontal ui-menu--primary" current="item-1-1-2" name="test-menu" role="menubar">
  <!---->
  <div class="ui-menu__inner">
    <ui-menu-item aria-label="Menu Item 1" aria-selected="" class="hydrated ui-menu-item ui-menu-item--selected" level="0" name="item-1" role="menuitem" selected="" tabindex="-1">
      <!---->
      <div class="ui-menu-item__item">
        <span class="ui-menu-item__label">
          <span class="ui-menu-item__label-slot"></span>
          <span class="ui-menu-item__label-text">
            Menu Item 1
          </span>
        </span>
        <ui-icon class="hydrated">
          <svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
            <path d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"></path>
          </svg>
        </ui-icon>
      </div>
      <div class="ui-menu-item__submenu ui-menu-item__submenu--below">
        <ui-menu-item aria-label="Menu Item 1-1" aria-selected="" class="hydrated ui-menu-item ui-menu-item--selected" level="1" name="item-1-1" role="menuitem" selected="" tabindex="-1">
          <!---->
          <div class="ui-menu-item__item">
            <span class="ui-menu-item__label">
              <span class="ui-menu-item__label-slot"></span>
              <span class="ui-menu-item__label-text">
                Menu Item 1-1
              </span>
            </span>
            <ui-icon class="hydrated">
              <svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
                <path d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"></path>
              </svg>
            </ui-icon>
          </div>
          <div class="ui-menu-item__submenu ui-menu-item__submenu--aside">
            <ui-menu-item aria-label="Menu Item 1-1-1" class="hydrated ui-menu-item" level="2" name="item-1-1-1" role="menuitem" tabindex="-1">
              <!---->
              <div class="ui-menu-item__item">
                <span class="ui-menu-item__label">
                  <span class="ui-menu-item__label-slot"></span>
                  <span class="ui-menu-item__label-text">
                    Menu Item 1-1-1
                  </span>
                </span>
              </div>
            </ui-menu-item>
            <ui-menu-item aria-label="Menu Item 1-1-2" aria-selected="" class="hydrated ui-menu-item ui-menu-item--selected" level="2" name="item-1-1-2" role="menuitem" selected="" tabindex="-1">
              <!---->
              <div class="ui-menu-item__item">
                <span class="ui-menu-item__label">
                  <span class="ui-menu-item__label-slot"></span>
                  <span class="ui-menu-item__label-text">
                    Menu Item 1-1-2
                  </span>
                </span>
              </div>
            </ui-menu-item>
          </div>
        </ui-menu-item>
      </div>
    </ui-menu-item>
  </div>
  <ui-menu-indicator class="hydrated ui-menu-indicator ui-menu-indicator--horizontal" style="display: block; transform: translateX(0px); width: 146px;"></ui-menu-indicator>
</ui-menu>
`;
